streaming/base/storage/upload.py

```diff
@@ -0,0 +1,355 @@
+
+    def _validate(self, out: Union[str, Tuple[str, str]]) -> None:
+        """Validate the `out` argument.
+
+        Args:
+            out (str | List[str]): Output dataset directory to save shard files.
+                1. If `out` is a local directory, shard files are saved locally.
+                2. If `out` is a remote directory, a random local temporary directory is created to
+                   cached the shard files and then the shard files are uploaded to a remote
+                   location. At the end, a temp directory is deleted once shards are uploaded.
+                3. If `out` is a list of `(local_dir, remote_dir)`, shard files are saved in the
+                   `local_dir` and also uploaded to a remote location.
+
+        Raises:
+            ValueError: Invalid number of `out` argument.
+            ValueError: Invalid Cloud provider prefix.
+        """
+        if isinstance(out, str):
+            obj = urllib.parse.urlparse(out)
+        else:
+            if len(out) != 2:
+                raise ValueError(''.join([
+                    f'Invalid `out` argument. It is either a string of local/remote directory ',
+                    'or a list of two strings with [local, remote].'
+                ]))
+            obj = urllib.parse.urlparse(out[1])
+        if obj.scheme not in WRITERS:
+            raise ValueError('Invalid Cloud provider prefix.')
+
+    def __init__(self,
+                 out: Union[str, Tuple[str, str]],
+                 keep_local: bool = False,
+                 progress_bar: bool = False) -> None:
+        """Initialize and validate local and remote path.
+
+        Args:
+            out (str | List[str]): Output dataset directory to save shard files.
+                1. If `out` is a local directory, shard files are saved locally.
+                2. If `out` is a remote directory, a random local temporary directory is created to
+                   cached the shard files and then the shard files are uploaded to a remote
+                   location. At the end, a temp directory is deleted once shards are uploaded.
+                3. If `out` is a list of `(local_dir, remote_dir)`, shard files are saved in the
+                   `local_dir` and also uploaded to a remote location.
+            keep_local (bool): If the dataset is uploaded, whether to keep the local dataset
+                shard file or remove it after uploading. Defaults to ``False``.
+            progress_bar (bool): Whether to display a progress bar while uploading to a cloud
+                provider. Defaults to ``False``.
+
+        Raises:
+            FileExistsError: Local directory must be empty.
+        """
+        self._validate(out)
+        self.keep_local = keep_local
+        self.progress_bar = progress_bar
+
+        if isinstance(out, str):
+            # It is a remote directory
+            if urllib.parse.urlparse(out).scheme:
```

Comment:
```suggestion
            if urllib.parse.urlparse(out).scheme != '':
```

Reply:
Addressed as part of https://github.com/mosaicml/streaming/pull/171/commits/622e74386e49cfd517641a01b340929bed47c675.